wire-grpc-client/src/main/java/com/squareup/wire/internal/grpc.kt

```diff
@@ -0,0 +1,137 @@
+import com.squareup.wire.GrpcMessageSource
+import com.squareup.wire.PipeDuplexRequestBody
+import com.squareup.wire.ProtoAdapter
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.channels.ReceiveChannel
+import kotlinx.coroutines.channels.SendChannel
+import kotlinx.coroutines.channels.consumeEach
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import okhttp3.Call
+import okhttp3.Callback
+import okhttp3.MediaType
+import okhttp3.MediaType.Companion.toMediaType
+import okhttp3.RequestBody
+import okhttp3.Response
+import okio.BufferedSink
+import java.io.IOException
+
+internal val APPLICATION_GRPC_MEDIA_TYPE: MediaType = "application/grpc".toMediaType()
+
+/** Returns a new request body that writes [onlyMessage]. */
+internal fun <S : Any> newRequestBody(
+  requestAdapter: ProtoAdapter<S>,
+  onlyMessage: S
+): RequestBody {
+  return object : RequestBody() {
+    override fun contentType() = APPLICATION_GRPC_MEDIA_TYPE
+
+    override fun writeTo(sink: BufferedSink) {
+      GrpcMessageSink(sink, requestAdapter, grpcEncoding = "identity").use {
+        it.write(onlyMessage)
+      }
+    }
+  }
+}
+
+/**
+ * Returns a new duplex request body that allows us to write request messages even after the
+ * response status, headers, and body have been received.
+ */
+internal fun newDuplexRequestBody(): PipeDuplexRequestBody {
+  return PipeDuplexRequestBody(APPLICATION_GRPC_MEDIA_TYPE, pipeMaxBufferSize = 1024 * 1024)
+}
+
+/** Writes messages to the request body. */
+internal fun <S : Any> PipeDuplexRequestBody.messageSink(
+  requestAdapter: ProtoAdapter<S>
+): GrpcMessageSink<S> {
+  return GrpcMessageSink(createSink(), requestAdapter, grpcEncoding = "identity")
+}
+
+/** Sends the response messages to the channel. */
+internal fun <R : Any> SendChannel<R>.readFromResponseBodyCallback(
+  responseAdapter: ProtoAdapter<R>
+): Callback {
+  return object : Callback {
+    override fun onFailure(call: Call, e: IOException) {
+      // Something broke. Kill the response channel.
+      close(e)
+    }
+
+    override fun onResponse(call: Call, response: Response) {
+      runBlocking {
+        response.use {
+          response.messageSource(responseAdapter).use { reader ->
+            while (true) {
+              val message = reader.read() ?: break
+              send(message)
+            }
+
+            close(response.grpcStatusToException())
+          }
+        }
+      }
+    }
+  }
+}
+
+/**
+ * Stream messages from the request channel to the request body stream. This means:
+ *
+ * 1. read a message (non blocking, suspending code)
+ * 2. write it to the stream (blocking)
+ * 3. repeat. We also have to wait for all 2s to end before closing the writer
+ */
+internal fun <S : Any> ReceiveChannel<S>.writeToRequestBody(
+  requestBody: PipeDuplexRequestBody,
+  requestAdapter: ProtoAdapter<S>
+) {
+  CoroutineScope(Dispatchers.IO).launch {
```

Comment:
I realize you copied this code, but this seems like a leak. This should be a suspend function and the caller should do something like
```kotlin
runBlocking {
  launch(IO) {
    channel.writeToRequestBody(.., ..)
  }
  somethingElse()
}
```